frontend/src/components/Forum/Forum.tsx

```diff
@@ -0,0 +1,123 @@
+
+    let mockData = Array.from(Array(10).keys()).map(idx => getMockThread(idx));
+   
+    const COLS = Array.from(Array(headers.length).keys());
+    const ROWS = Array.from(Array(mockData.length).keys());
+
+    const [isFavourite, setFavourite] = useState(mockData.map(elem => elem.isFavourite));
+
+    function toggleFavourite(idx: number){
+        mockData[idx].isFavourite = !mockData[idx].isFavourite;
+        setFavourite(isFavourite.map((element, currIdx) => currIdx===idx ? !element : element));
+    };
+
+    function getData(rowNum: number, colNum: number){
+        var data;
+        switch(colNum){
+            case 0: {
+                data = mockData[rowNum].title;
+                break;
+            }
+            case 1: {
+                data = mockData[rowNum].forumPosts.length;
+                break;
+            }
+            case 2: {
+                data = mockData[rowNum].creator.username;
+                break;
+            }
+            case 3: {
+                let date = mockData[rowNum].dateCreated;
+                data = date.getFullYear()+"-"+(date.getMonth()+1)+"-"+date.getDate();
+                break;
+            }
+            default: {
+                data = "---";
+                break;
+            }
+                
+        }
+        return data;
+    }
+
+    function getHeaderRow(){
+        return(
+            <ForumRow  className="m-0">
+                {COLS.map(idx => (<ForumCol key={idx} className={classes[idx]}>
+                                    <ForumHeader key={idx} className={classes[idx]}>{headers[idx]}</ForumHeader>
+                                </ForumCol>))}
+            </ForumRow>
+        )
+    }
+
+    function getTableElem(rowNum: number, colNum: number) {
+        let data = getData(rowNum, colNum);
+        let last = (colNum === (headers.length-1));
+        return (
+            last === false ? 
+                <ForumTile key={rowNum+'.'+colNum} className='not-last'>{data}</ForumTile>  //className={name}
+                : 
+                <ForumTile key={rowNum+'.'+colNum}>
+                    <OpenButton>Otwórz</OpenButton>
+                    <Star icon = {faStarSolid} 
+                        className={isFavourite[rowNum] ? 'starred': 'unstarred'} 
+                        onClick={() => toggleFavourite(rowNum)} 
+                    />
+                </ForumTile>
+        )
+    }
+
+    function getThread(rowNum: number){
+        return(
+            <ForumRow key={rowNum} className="m-0">
+                {COLS.map(colNum => (
+                    <ForumCol key={rowNum+'.'+colNum} className={classes[colNum]}>
+                        {getTableElem(rowNum, colNum)}
+                    </ForumCol>
+                ))}
+            </ForumRow>
+        )
+    }
+
+    return (
+        <ForumContainer>
+            <Row className="mt-5">
+                    {getHeaderRow()}
+                    {ROWS.map(idx => getThread(idx))}
```

Comment:
Wydaje mi się że wciąż jest to trochę przekomplikowane, można po prostu zrobić 

```tsx
mockData.map(thread => {
// ...tutaj kod który bierze wszelkie dane z thread i robi z nich pojedynczy wiersz, np cos takiego jak
return (
   <ForumRow key={thread.id}>
      <ForumTile className={nazwa_klasy_dla_tytulu}>{thread.title}</ForumTile>
      <ForumTile className={nazwa_klasy_dla_liczby_odpowiedzi}>{thread.forumPosts.length}</ForumTile>
            ... itd
       <ForumTile ...>
           <OpenButton onClick={() => otworz_karte_z_watkiem(thread)}/>
           <Star onClick={() => funkcja_co_robi: thread.favaourite = !thread.favourite}>
           ...
       <ForumTile>
   </ForumRow>
)
}
```
Powinny wtedy zniknąć wszelkie problemy z indexami i zarządzeniem tym wszystkim. To z toggleFavourite może Ci po takiej edycji nie działać bo co chwile generujesz te wątki od nowa, jak już będą dane z bazy to tak się nie będzie działo, na ten moment można nawet przenieść te generacje wątków poza komponent (na zewnątrz w tym samym pliku)